### DISCUSSION TITLE
Why a matrix step will be canceled if another one failed

### DISCUSSION THREAD

**Discussion by zaherg** [community/community#27192] [Actions]
<p>Hey,</p>
<p>I have a small PHP Job as the following</p>
<pre><code>phpunit:
    name: PHPUnit
    runs-on: ubuntu-latest
    services:
      mysql:
        image: mysql:5.7
        env:
          MYSQL_ALLOW_EMPTY_PASSWORD: false
          MYSQL_ROOT_PASSWORD: password
          MYSQL_DATABASE: laravel
        ports:
          - 3306
        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
    strategy:
      matrix:
        php: [7.2, 7.3]
    steps:
      - uses: actions/checkout@v1
      - name: Install Composer dependencies
        run: composer install --no-progress --no-suggest --prefer-dist --optimize-autoloader
      - name: Prepare the application
        run: |
          cp .env.example .env
          php artisan key:generate
      - name: Run Migration
        run: php${{ matrix.php }} artisan migrate -v
        env:
          DB_PASSWORD: password
</code></pre>
<p>If the tests failed in PHP 7.2 the workflow will not process any tests in PHP 7.3</p>
<p><img src="https://user-images.githubusercontent.com/27624/181102202-7b1d5f82-3984-4214-85de-8ad368fa7bf8.png" alt="Screen Shot 2019-09-07 at 1.54.30 PM.png" data-base62-sha1="8RFwJW0gE8HnMhu21URJY8A0PBt" width="347" height="331"></p>
<p>This is an example of the issue <a href="https://github.com/linuxjuggler/laravel-mysql-test/commit/18c405d62df99a5573145b48ac1888bcd4b743f2/checks" rel="nofollow noopener">https://github.com/linuxjuggler/laravel-mysql-test/commit/18c405d62df99a5573145b48ac1888bcd4b743f2/checks</a></p>

**Comment by Legion2** [comment]
<p>This is a fail-fast behavior, if one job fails the workflow fails and all jobs are discarded. If you run tests and one fails you don’t need to run more tests, because you already know there is a problem.</p>

**Comment by zaherg** [comment]
<p>This does not make any sense, especially that am testing the code with two different PHP version.</p>
<p>Everywhere, it’s common for your code to work on PHP 7.2 for example and fail on 7.3 or vice-versa.</p>
<p>The same when your code, for example, pass a test on Mac but fail on Linux, it means that your mac build has a problem not all your code.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/4/49dc9f35e86eab8924274290ef5c36d269395c71.png" data-download-href="/uploads/short-url/axpxcUTbT8mDlwWCBtRnxpi80sF.png?dl=1" title="Screen Shot 2019-09-07 at 5.29.11 PM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/27624/181102205-370819c4-32f5-4b35-9c4c-5ae422ef9d7a.png" alt="Screen Shot 2019-09-07 at 5.29.11 PM.png" data-base62-sha1="axpxcUTbT8mDlwWCBtRnxpi80sF" width="690" height="248" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/4/49dc9f35e86eab8924274290ef5c36d269395c71_2_690x248.png, https://aws1.discourse-cdn.com/github/optimized/2X/4/49dc9f35e86eab8924274290ef5c36d269395c71_2_1035x372.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/4/49dc9f35e86eab8924274290ef5c36d269395c71.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/4/49dc9f35e86eab8924274290ef5c36d269395c71_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-09-07 at 5.29.11 PM.png</span><span class="informations">1095×395 38.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>So when my code fails in 7.2, it means I have a problem when using PHP 7.2 it does not mean by default that all my code will fail no matter which version of PHP am using.</p>
<p>Your code should be tested for all the matrixes, not just one of them!!!</p>

**Comment by Legion2** [comment, marked as answer]
<p>You can turn off fail-fast behavior of your workflow see: <a href="https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast" rel="nofollow noopener">https://help.github.com/en/articles/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast</a></p>

**Comment by ZacSweers** [comment]
<p>It would be great if we could specify matrix configurations that should be allowed to fail. This is useful for a few cases, such as integration testing against new versions of some base framework, WIP support for certain environments, etc. It’s possible to do this by manually controlling success()/fail() checks, but it’s not as nice as being able to define this at the matrix level. Travis’ documentation for their equivalent: <a href="https://docs.travis-ci.com/user/customizing-the-build/#rows-that-are-allowed-to-fail" rel="nofollow noopener">https://docs.travis-ci.com/user/customizing-the-build/#rows-that-are-allowed-to-fail</a></p>

**Comment by zaherg** [comment]
<p>I think (as I didn’t test it yet) you can use <a href="https://help.github.com/en/articles/contexts-and-expression-syntax-for-github-actions" title="Contexts and expression syntax for GitHub Actions" rel="nofollow noopener">Contexts and expression syntax for GitHub Actions</a> for that</p>

**Comment by psafont** [comment]
<p>I was able to achieve this with continue-on-error: <a href="https://help.github.com/en/github/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error" rel="nofollow noopener">https://help.github.com/en/github/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error</a></p>
<p>Add an expression to the field which reads from matrix variables to ignore failures, for example:</p>
<pre><code>jobs:
build:
runs-on: ubuntu-18.04
strategy:
matrix:
ocaml: ["4.07", "4.08", "4.09"]
steps:
- name: Checkout
uses: actions/checkout@v1
- name: Build
run: make
env:
OCAML_VERSION: ${{ matrix.ocaml }}
continue-on-error: ${{matrix.ocaml == '4.09' }}
</code></pre>
